src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -249,70 +205,135 @@ protected virtual string GenerateUniqueId()
-                if (!attributeValueRecorded)
+        private void RenderAttributes(
+            IList<KeyValuePair<string, Chunk>> chunkAttributes,
+            IEnumerable<TagHelperDescriptor> tagHelperDescriptors)
+        {
+            var renderedBoundAttributeNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            // Go through the HTML attributes in source order, assigning to properties or indexers or adding to
+            // TagHelperExecutionContext's HTML attributes as we go.
+            foreach (var attribute in chunkAttributes)
+            {
+                var associatedDescriptors = tagHelperDescriptors.Where(descriptor =>
+                    descriptor.Attributes.Any(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key)));
+
+                // Bound attributes will have associated descriptors.
+                if (associatedDescriptors.Any())
                 {
-                    // We only need to create attribute values once per HTML element (not once per tag helper).
-                    // We're saving the value accessor so we can retrieve it later if there are more tag
-                    // helpers that need the value.
-                    htmlAttributeValues.Add(attributeName, valueAccessor);
-
-                    // Bufferable attributes are attributes that can have Razor code inside of them. Such
-                    // attributes have string values and may be calculated using a temporary TextWriter or other
-                    // buffer.
-                    var bufferableAttribute = attributeDescriptor.IsStringProperty;
-
-                    RenderNewAttributeValueAssignment(
-                        attributeDescriptor,
-                        bufferableAttribute,
-                        attributeValueChunk,
-                        valueAccessor);
-
-                    // Execution contexts are a runtime feature.
-                    if (_designTimeMode)
+                    if (!renderedBoundAttributeNames.Add(attribute.Key))
                     {
+                        // We only need to create attribute values once per HTML element (not once per tag helper).
                         continue;
                     }
 
-                    // We need to inform the context of the attribute value.
-                    _writer
-                        .WriteStartInstanceMethodInvocation(
-                            ExecutionContextVariableName,
-                            _tagHelperContext.ExecutionContextAddTagHelperAttributeMethodName)
-                        .WriteStringLiteral(attributeName)
-                        .WriteParameterSeparator()
-                        .Write(valueAccessor)
-                        .WriteEndMethodInvocation();
+                    // We need to capture the tag helper's property value accessor so we can retrieve it later
+                    // (if we need to).
```

Comment:
`(if we need to)` worked only because more detail was provided earlier in the old code. Change to `... if there are more tag helpers that need the value`.